File: java/leetcode/MaxSumRectangle.java
//Leetcode problem 363 Max Sum of Rectangle No Larger Than K
//Solution written by Xuqiang Fang on 12 July, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
import java.util.Collections;
class Solution{
    public int maxSumSubmatrix(int[][] matrix, int k) {
        int m = Math.min(matrix.length, matrix[0].length);   
        int n = Math.max(matrix.length, matrix[0].length);
        int ans = Integer.MIN_VALUE; 
        for(int i=0; i<m; ++i){
            int[] sum = new int[n];
            for(int j=i; j<m; ++j){
                for(int l=0; l<n; ++l){
                    if(m == l){
                        sum[l] += matrix[j][l];
                    }
                    else{
                        sum[l] += matrix[l][j];
                    }
                } 
                List<Integer> list = new ArrayList<>();
                int curr = 0;
                list.add(0);
                for(int x=0; x<n; ++x){
                    curr += sum[x];
                    int index = Collections.binarySearch(list, curr-k);
                    if(index < 0){
                        index = -1 - index; 
                    }
                    if(index != list.size()){
                        ans = Math.max(ans, curr-list.get(index));
                    }
                    list.add(index, curr);
                }
            }
        }
        return ans;
    }
}

public class MaxSumRectangle{
	public static void main(String[] args){
		Solution s = new Solution();
        int[][] matrix = {{1,0,1},{0,-2,3}};
        System.out.println(s.maxSumSubmatrix(matrix, 2));
	}
}

File: python/leetcode/MaxSumRectangle.py
'''
Leetcode problem No 363 Max Sum of Rectangle No Larger Than K
Solution written by Xuqiang Fang on 12 July, 2018
'''
from bisect import bisect_left, insort
class Solution(object):
    def maxSumSubmatrix(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        if not matrix:
            return 0
        m = min(len(matrix), len(matrix[0]))
        n = max(len(matrix), len(matrix[0]))
        ans = float('-inf')
        for i in range(m):
            s = [0] * n
            for j in range(i, m):
                for l in range(n):
                    s[l] += matrix[j][l] if m == len(matrix) else  matrix[l][j]
                accum_set, accum = [0], 0
                for t in s:
                    accum += t
                    it = bisect_left(accum_set, accum-k)
                    print(accum_set)
                    print(it)
                    if it != len(accum_set):
                        ans = max(ans, accum-accum_set[it])
                    insort(accum_set, accum)
        return ans

def main():
    s = Solution()
    matrix = [[1,0,1],[0,-2,3]]
    print(s.maxSumSubmatrix(matrix, 2))

if __name__ == '__main__':
    main()
